pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandler.java

```diff
@@ -677,4 +836,168 @@ private void updateMetadataProperties(File indexDir, Map<String, String> metadat
+      // lengthOfLongestEntry not available for dict columns. Should be computed by reading all values in dictionary.
+      IndexCreationContext.Builder builder =
+          IndexCreationContext.builder().withIndexDir(indexDir).withColumnMetadata(existingColMetadata);
+      builder.withDictionary(false);
+
+      if (existingColMetadata.isSingleValue()) {
+        int lengthOfLongestEntry = getLengthOfLongestEntryFromDictionary(column, reader, dictionary);
+        builder.withLengthOfLongestEntry(lengthOfLongestEntry);
+      } else {
+        FieldSpec.DataType dataType = existingColMetadata.getDataType();
+        boolean isFixedWidth = dataType.getStoredType().isFixedWidth();
+
+        if (!isFixedWidth) {
+          // For variable length stored types, maxRowLengthInBytes is required to create the forwardIndexCreator.
+          // This can only be determined by reading the entire MV forward index.
+          int maxRowLength = getMaxRowLengthForMVColumn(column, reader, dictionary);
+          builder.withMaxRowLengthInBytes(maxRowLength);
+        }
+      }
+
+      Map<String, ChunkCompressionType> compressionConfigs = _indexLoadingConfig.getCompressionConfigs();
+      ChunkCompressionType compressionType;
+      if (compressionConfigs.containsKey(column)) {
+        compressionType = compressionConfigs.get(column);
+      } else {
+        compressionType = SegmentColumnarIndexCreator.getDefaultCompressionType(existingColMetadata.getFieldType());
+      }
+
+      IndexCreationContext.Forward context =
+          builder.build().forForwardIndex(compressionType, _indexLoadingConfig.getColumnProperties());
+      try (ForwardIndexCreator creator = indexCreatorProvider.newForwardIndexCreator(context)) {
+        int numDocs = existingColMetadata.getTotalDocs();
+        forwardIndexRewriteHelper(column, existingColMetadata, reader, creator, numDocs, null, dictionary);
+      }
+    }
+  }
+
+  private int getMaxRowLengthForMVColumn(String column, ForwardIndexReader reader, Dictionary dictionary)
+      throws Exception {
+    ColumnMetadata existingColMetadata = _segmentMetadata.getColumnMetadataFor(column);
+    AbstractColumnStatisticsCollector statsCollector =
+        getStatsCollector(column, dictionary.getValueType().getStoredType());
+    PinotSegmentColumnReader columnReader =
+        new PinotSegmentColumnReader(reader, dictionary, null, existingColMetadata.getMaxNumberOfMultiValues());
+    int numDocs = existingColMetadata.getTotalDocs();
+
+    Preconditions.checkState(!existingColMetadata.getDataType().getStoredType().isFixedWidth(),
+        "Column " + column + "is not a fixed width column.");
+    Preconditions.checkState(!existingColMetadata.isSingleValue(), "Column " + column + "is not MV.");
+
+    for (int i = 0; i < numDocs; i++) {
+      Object obj = columnReader.getValue(i);
+      statsCollector.collect(obj);
+    }
+
+    statsCollector.seal();
```

Comment:
Same here. I don't think we need to `StatsCollector` unless you are doing it for some efficiency reasons.